management/sites/sitesController.js

```diff
@@ -0,0 +1,311 @@
+
+		$scope.addShiftCancelButtonHandler = () => {
+			$scope.addShiftInformation = {};
+			$scope.addShiftButtonClicked = false;
+		};
+
+		$scope.addShiftStartTimeChanged = (newStartTime) => {
+			// Change end time options to only be the ones past the start time
+			const period = newStartTime.slice(-2);
+			const hourString = newStartTime.split(':')[0];
+			const minuteString = newStartTime.split(':')[1];
+
+			let hour = parseInt(hourString);
+			if (period === 'AM' && hour === 12) hour = 0;
+			if (period === 'PM' && hour !== 12) hour += 12;
+			const minute = parseInt(minuteString);
+			const time = hour * 60 + minute + TIME_INTERVAL; // + TIME_INTERVAL so the same time can't be selected
+
+			$scope.addShiftEndTimeOptions = generateTimes(TIME_INTERVAL, time).map(time => time.timeString);
+		};
+
+		function isShiftDateEqualToDate(shiftDate, date) {
+			return new Date(new Date(shiftDate).toDateString()).getTime() === new Date(date.toDateString()).getTime();
+		};
+
+		$scope.addAppointmentTimeButtonHandler = () => {
+			$scope.initializeDatePicker('addAppointmentTimeDateInput', (date) => {
+				$scope.addAppointmentTimeInformation.selectedDate = date;
+				$scope.updateAddAppointmentTimeScheduledTimeOptions();
+				$scope.$apply();
+			}, (date) => {
+				const dateInPast = isDateInPast(date);
+				const anyShiftsOnThisDay = $scope.siteInformation.shifts
+					.some(shift => isShiftDateEqualToDate(shift.startTime, date));
+				return dateInPast || !anyShiftsOnThisDay;
+			});
+			$scope.addAppointmentTimeButtonClicked = true;
+		};
+
+		$scope.addAppointmentTimeApproximateLengthInMinutesChanged = (newApproximateLengthInMinutes) => {
+			$scope.addAppointmentTimeInformation.approximateLengthInMinutes = newApproximateLengthInMinutes;
+			$scope.updateAddAppointmentTimeScheduledTimeOptions();
+		};
+ 
+		$scope.updateAddAppointmentTimeScheduledTimeOptions = () => {
+			const timeInterval = $scope.addAppointmentTimeInformation.approximateLengthInMinutes;
+			if (timeInterval == null || timeInterval <= 0) {
+				$scope.addAppointmentTimeScheduledTimeOptions = [];
+				return;
+			}
+
+			const date = $scope.addAppointmentTimeInformation.selectedDate;
+			const shiftsOnThisDay = $scope.siteInformation.shifts
+				.filter(shift => isShiftDateEqualToDate(shift.startTime, new Date(date)));
+
+			const times = $scope.getAvailableAppointmentTimes(shiftsOnThisDay, timeInterval);
+			$scope.addAppointmentTimeScheduledTimeOptions = times.map(time => time.timeString);
+		};
+
+		$scope.getAvailableAppointmentTimes = (shifts, timeInterval) => {
+			const getTime = (time) => {
+				const dateTime = new Date(time);
+				return dateTime.getHours() * 60 + dateTime.getMinutes();
+			};
+
+			let times = [];
+
+			// Get the possible times within each shift
+			for (const shift of shifts) {
+				const startTime = getTime(shift.startTime);
+				const endTime = getTime(shift.endTime);
+
+				const generatedTimes = generateTimes(timeInterval, startTime, endTime);
```

Comment:
I think for the third parameter in `generateTimes` really depends on if the times need to fall completely within the start and end times or if they just need to start in that range. For the appointment times, we want those to fall entirely within the shift. I think we should use `endTime - timeInterval` instead of just `endTime`. It correctly throws an error, but I feel like we should just not show that option.

Reply:
It wasn't enough to use `endTime - timeInterval` on this because if a shift time went from 2pm-4pm and we have appointments that are 60 minutes long, then we wanted to have both a 2pm and 3pm appointment times. 

So what I did was use `endTime - timeInterval` here and then have a `endTimeInclusive` parameter that defaults to false in that `generateTimes` method, and when it is set to true, the method will generate a time for the `endTime` part as well.. Thoughts on if this is a good approach?

Another option that would've worked was using `endTime - timeInterval + 1` instead of any of the other changes :p

Reply:
Ah. Makes sense. The changes look good.